libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder;
+
+	builder.m_contract = &_contract;
+
+	builder.m_graph = make_unique<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	builder.m_currentNode = SpecialNode::EntryCreation;
+	builder.m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
```

Comment:
This is probably not yet *exactly* the right order, but I'd leave it at this for this PR - but we should update the comment in the header, stating that the initialization order in the creation is not exactly preserved - then we can refine that when this is actually used for stuff like unreachable code when we can actually add tests for it.

Reply:
@ekpyron Can you expand on that? Do you mean that the lines to functions called with initializing state variables should be coming from specific constructors rather than from `EntryCreation`?

Reply:
Quite an old comment you dug up there, I almost didn't find it after seeing the github notification :-).
Ideally, the connections in the graph should correspond exactly to what codegen does during construction.
So basically following ``IRGenerator::generateImplicitConstructors`` you'd have an edge from the entry to an implicit constructor, from that edges to whatever function calls the base arguments involve, to the next constructor in the hierarchy and to whatever statevariable initilization involves. Then only the next constructor has edges to the next base arguments and so on... I'd check if that's the case right now or not, but github has a really hard time dealing with this PR apparently :-D....

Reply:
Well, I went through them all to see what's still relevant :)

ok, so yeah, probably better to do that in a separate PR and keep work in this one to a minimum.